Fix reading from cache in rz_io_nread_at().

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).
- [ ] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request.
If you have used AI tools to generate these code changes, please disclose software used, model name. -->

    
    The cache read must read within the mapped region and not beyond.
    Otherwise it would go against the documentation of rz_io_nread_at().
    rz_io_cache_read() internally doesn't know about map boundaries so
    the length parameter passed to it in rz_io_nread_at()
    should only read 'ret' number of bytes.
    
    To give somewhat more introspective what rz_io_cache_read() does,
    the return value meaning was changed to success/failure
    and an out parameter was added to get the number of bytes read.


**Test plan**

<!-- THIS IS A HARD REQUIREMENT FOR NEW CONTRIBUTORS! -->
<!-- Please refer to CONTRIBUTING.md for more details. -->

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

All green

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

notxvilka reviewed (approved):
Good catch!

Rot127 commented:
~~Something is broken with the cache :roll_eyes:~~
~~It doesn't read the amount of data requested.~~

Never mind. Just non existing documentation.

notxvilka commented:
```
Mismatched function annotation for rz_io_cache_read at <rizin/librz/include/rz_io.h:407:12> : RZ_BORROW (was None at <rizin/librz/io/io_cache.c:154:12>
Mismatched function annotation for rz_io_cache_read at <rizin/librz/include/rz_io.h:407:12> : RZ_NONNULL (was None at <rizin/librz/io/io_cache.c:154:12>
```

Rot127 committed:
Fix reading from cache in rz_io_nread_at().

The cache read must read within the mapped region and not beyond.
Otherwise it would go against the documentation of rz_io_nread_at().
rz_io_cache_read() internally doesn't know about map boundaries so
the length parameter passed to it in rz_io_nread_at()
should only read 'ret' number of bytes.

To give somewhat more introspective what rz_io_cache_read() does,
the return value meaning was changed to success/failure
and an out parameter was added to get the number of bytes read.

Rot127 commented:
Will fix it soon. I still struggle to find a consistent way of making the changes.

Rot127 commented:
Ok, new try.
